### DISCUSSION TITLE
Context management restores soft-se root keys?

### DISCUSSION THREAD

**Discussion by brocaar** [Lora-net/LoRaMac-node#1067] [Q&A]
It looks like when `CONTEXT_MANAGEMENT_ENABLED` is enabled, together with the soft-se option, that changing the keys in `se-identity.h` has no effect. I noticed this after updating the `LORAWAN_JOIN_EUI` and `NWK_KEY`. After (re)flashing, the device kept using the old `LORAWAN_JOIN_EUI` (and I assume as well the `NWK_KEY`, as the device failed to join).

I tried removing the `build` folder, `make clean` etc... but none of my changes had any effect. Then I noticed the `###### =========== CTXS RESTORED =========== ######` message. After disabling `CONTEXT_MANAGEMENT_ENABLED`, rebuild and reflash, the device succesfully joined using the new `LORAWAN_JOIN_EUI` and `NWK_KEY`.

**Comment by mluis1** [comment]
A few questions:
- Which version are you using?
- Which platform?
- Have you tried to make a mass-erase of the MCU Flash memory?
   The NVM data memory doesn't not get erased by updating the firmware.

If you are using a version newer or equal to v4.4.7 (LoRaWAN 1.0.3), v.4.5.1 (LoRaWAN 1.0.4) or master (1.0.4)/develop(1.0.4/1.1.1) branch you have the possibility to erase the NVM data through the serial interface by pressing ESC then shift N keys.

**Comment by mluis1** [comment]
If you don't mind I move this issue to the [Discussions](https://github.com/Lora-net/LoRaMac-node/discussions) tab.

**Comment by brocaar** [comment]
Version: d02b2f86000a838a1c723757d0f0ecf2327555a7
Platform: B-L072Z-LRWAN1

I can still reproduce this:

1. I create a firmware with DevEUI / JoinEUI / AppKey A
2. I flash A to the device
3. I let the device join
4. Device joins with A
5. I create a new firmware with DevEUI / JoinEUI / AppKey B
6. I flash B to the device
7. I power-cycle the device
8. Device joins with A
9. I power-cycle the device
10. Device joins with A
11. ....
9. ESC & shift+N + power-cycle
10. Device joins with B

**Comment by brocaar** [comment]
These are the local changes that I have made:

```diff
$ git diff
diff --git a/src/apps/LoRaMac/periodic-uplink-lpp/B-L072Z-LRWAN1/main.c b/src/apps/LoRaMac/periodic-uplink-lpp/B-L072Z-LRWAN1/main.c
index 89bb63bd..031090c4 100644
--- a/src/apps/LoRaMac/periodic-uplink-lpp/B-L072Z-LRWAN1/main.c
+++ b/src/apps/LoRaMac/periodic-uplink-lpp/B-L072Z-LRWAN1/main.c
@@ -73,7 +73,7 @@
  *
  * \remark Please note that LORAWAN_DEFAULT_DATARATE is used only when ADR is disabled 
  */
-#define LORAWAN_DEFAULT_DATARATE                    DR_0
+#define LORAWAN_DEFAULT_DATARATE                    DR_5
 
 /*!
  * LoRaWAN confirmed messages
diff --git a/src/peripherals/soft-se/se-identity.h b/src/peripherals/soft-se/se-identity.h
index 872167b4..c9abc932 100644
--- a/src/peripherals/soft-se/se-identity.h
+++ b/src/peripherals/soft-se/se-identity.h
@@ -66,17 +66,17 @@ extern "C" {
  * When set to 1 DevEui is LORAWAN_DEVICE_EUI
  * When set to 0 DevEui is automatically set with a value provided by MCU platform
  */
-#define STATIC_DEVICE_EUI                                  0
+#define STATIC_DEVICE_EUI                                  1
 
 /*!
  * end-device IEEE EUI (big endian)
  */
-#define LORAWAN_DEVICE_EUI                                 { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 }
+#define LORAWAN_DEVICE_EUI                                 { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x04 }
 
 /*!
  * App/Join server IEEE EUI (big endian)
  */
-#define LORAWAN_JOIN_EUI                                   { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 }
+#define LORAWAN_JOIN_EUI                                   { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x04 }
 
 /*!
  * Secure-element pin
@@ -112,8 +112,8 @@ extern "C" {
              * WARNING: FOR 1.0.x DEVICES IT IS THE \ref LORAWAN_APP_KEY                                            \
              */                                                                                                     \
             .KeyID    = NWK_KEY,                                                                                    \
-            .KeyValue = { 0x2B, 0x7E, 0x15, 0x16, 0x28, 0xAE, 0xD2, 0xA6, 0xAB, 0xF7, 0x15, 0x88, 0x09, 0xCF, 0x4F, \
-                          0x3C },                                                                                   \
+            .KeyValue = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, \
+                          0x04 },                                                                                   \
         },                                                                                                          \
         {                                                                                                           \
             /*!
```

**Comment by MedadRufus** [comment]
Only ESC & shift+N + power-cycle will wipe the old context data from the EEPROM. So if you want to change the soft-se keys, flash the new keys, then run ESC & shift+N + power-cycle.

**Reply by MedadRufus** [reply]
Another option is to manually wipe the EEPROM.

**Reply by mluis1** [reply]
@brocaar 

As said by @MedadRufus in your step list once you have executed step 12 the end-device erases the EEPROM data and then at step 13 it uses your new keys.

The EEPROM data is only erased once the MCU has received the `ESC` + `N` sequence. Then by reseting/power cycling the MCU it will notice that the EEPROM data is no more valid which makes the end-device to re-initialize everything.

https://github.com/Lora-net/LoRaMac-node/blob/d02b2f86000a838a1c723757d0f0ecf2327555a7/src/apps/LoRaMac/common/LmHandler/LmHandler.c#L256-L269

If `nbNvmData` is greater than 0 the firmware restores the EEPROM data. In case `nbNvmData` equals 0 the firmware initializes everything.

**Reply by brocaar** [reply]
Maybe this (ESC & <shift> + n) is something to add to the documentation? (https://github.com/Lora-net/LoRaMac-node/blob/master/README.md#soft-se)

> In order to update the end-device identity (DevEUI, JoinEUI and AES128 keys) one must update the se-identity.h file located under ./src/peripherals/soft-se/ directory.

That is what I did and not being aware that this state is not only stored inside the firmware binary but also persisted somewhere else in memory, it took me quite some time to figure out what was going on.

**Reply by mluis1** [reply]
You are right @brocaar. 
We did not had the time yet to update the documentation.
The non-volatile storage is pretty new (introduced with LoRaWAN 1.0.4 implementation) and users could use either MCU internal memory or an external non volatile memory.

**Reply by mluis1** [reply]
I have added a new chapter to the [README.md](https://github.com/Lora-net/LoRaMac-node/blob/master/README.md) file describing how to reset the NVM contents.
[Serial console NVM management](https://github.com/Lora-net/LoRaMac-node#serial-console-nvm-management)

**Reply by MedadRufus** [reply]
I just got caught by this requirement again today!

**Comment by shabaz-git** [comment]
Hi,
The same problem I am also facing. I am using fit iot-lab hardware (**b-l072z-lrwan1** module). 

When I observed this behavior I tried with erasing the NVM in the start of the application using `NvmDataMgmtFactoryReset()` . 
But this led into a late join response because of the DevNonce restriction. Is there any thing other than this, we can do to flash new eui's to same hardware.

To solve this can't we take a decision about the restoring the nvm data after comparing it with the data provided in the "_se-identity.h_" .
Is such implementation is already there?If this is not possible or not a good option, why?

Is there any way I can pass ESC + N.

My project is having a continuous power supply and for this I want to store the data in NVM only when I intended to power off the node. Is there any way to achieve this?

**Comment by bwesen** [comment]
I couldn't get soft-se to work either (latest stack). It seems there is a problem in the LmHandlerInit code, it first calls LoRaMacInitialization which does a *wipe* of the soft-se NVM, then it proceeds to 

```
    mibReq.Type = MIB_JOIN_EUI;
    LoRaMacMibGetRequestConfirm( &mibReq );
    memcpy1( CommissioningParams.JoinEui, mibReq.Param.JoinEui, 8 );
```

which, of course, then reads out 0's and puts in CommissioningParams.JoinEui no matter what.

What I really wanted to do was to use   SecureElementSetJoinEui(eui); to setup the Join EUI before LmHandlerInit but that won't work as that function re-initializes the soft-se module and wipes whatever you wrote there first.

I ended up having to monkeypatch a callback inside LmHandlerInit into my main code, which does the SetJoinEui there...
